### PR TITLE
Unicode conversion

### PR DESCRIPTION
**Pull request by hansenyy**
The Utf16ToUtf8 functions in UnicodeConversion.h is implemented using a utf16 to utf8 conversion function written by Facebook. Said Facebook function does some weird things to conform with JSC's encoding and has been removed from Facebook/react-native since the introduction of JSCRuntime. This PR does the following:

1. Reimplement the Utf16ToUtf8 functions in UnicodeConversion.h using WideCharToMultiByte.
2. Introduce a new exception class UnicodeConversionException, which derives from std::runtime_error. Have unicode conversion functions throw UnicodeConversionExceptions instead of std::runtime_errors.
3. Add a few more unicode conversion function overloads for convenience.
4. Rename the namespace UnicodeConversion to unicode.
5. Rename UnicodeConversion.h/cpp to unicode.h/.cpp
6. Rename Utf8ToUtf16 and Utf16ToUtf8 to utf8ToUtf16 and utf16ToUtf8 to conform with the naming convention in the rest of the repo.
7. As part of the reimplementation of Utf16ToUtf8, add two helper functions under ReactWindowsCore/utilities.h.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/2445)

### CODE REVIEW

File: vnext/ReactWindowsCore/utilities.h [hunk 1]
```diff
@@ -0,0 +1,47 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
```
Comment (hansenyy) [inline vnext/ReactWindowsCore/utilities.h]: I am aware that there is already a Utils.h file under ReactWindowsCore. However, looking at Utils.h, I feel that it should be renamed into Url.h. Any thoughts?

File: vnext/ReactWindowsCore/utilities.h [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+#pragma once
+#include <type_traits>
+#include <utility>
+
+namespace facebook {
+namespace react {
+namespace utilities {
```
Comment (hansenyy) [inline vnext/ReactWindowsCore/utilities.h]: Tests for functions under this namespace will be checked in as a separate pull request.

File: vnext/ReactWindowsCore/utilities.h [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    sizeof(TResultType) == sizeof(TOriginalType),
+    "checkedReinterpretCast can only be used to cast from T1* to T2*, where "
+    "T1 and T2 are integral types of the same size.");
+
+  return reinterpret_cast<TResultTypePtr>(p);
+}
+
+// A compile time function that deduces the size of an array.
+template<typename T, std::size_t N>
+constexpr std::size_t arraySize(T(&)[N]) noexcept
```
Comment (hansenyy) [inline vnext/ReactWindowsCore/utilities.h]: >arraySize [](start = 22, length = 9)

This function is not yet used anywhere in the code base. However, a similar function exists in the ChakraJsiRuntime implementations. This function will also be used by some unicode conversion tests that I plan on checking in later as a separate PR.

File: vnext/Desktop.DLL/react-native-win32.x64.def [hunk 4]
```diff
@@ -48,8 +48,17 @@ EXPORTS
 ?toJson@folly@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@AEBUdynamic@1@@Z
 ?typeName@dynamic@folly@@QEBAPEBDXZ
 ?updateProperties@ShadowNode@react@facebook@@UEAAX$$QEBUdynamic@folly@@@Z
-?Utf16ToUtf8@UnicodeConversion@react@facebook@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@AEBV?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@5@@Z
-?Utf8ToUtf16@UnicodeConversion@react@facebook@@YA?AV?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@5@@Z
+
+?utf8ToUtf16@unicode@react@facebook@@YA?AV?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@PEBD_K@Z
```
Comment (acoates-ms) [inline vnext/Desktop.DLL/react-native-win32.x64.def]: Lets not expand our dll boundary unless we need to.

Comment (acoates-ms) [inline vnext/Desktop.DLL/react-native-win32.x64.def]: Lets not expand our dll boundary unless we need to.
